modules/ggl-http/src/aws_sigv4.c

```diff
@@ -0,0 +1,348 @@
+// aws-greengrass-lite - AWS IoT Greengrass runtime for constrained devices
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "aws_sigv4.h"
+#include "ggl/http.h"
+#include "sigv4.h"
+#include <assert.h>
+#include <ggl/alloc.h>
+#include <ggl/buffer.h>
+#include <ggl/error.h>
+#include <ggl/log.h>
+#include <ggl/vector.h>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/ssl.h>
+#include <openssl/types.h>
+#include <openssl/x509.h>
+#include <string.h>
+#include <sys/time.h>
+#include <stdint.h>
+
+static int32_t hash_init(void *ctx) {
+    // Initialize OpenSSL digest
+    int32_t ret;
+
+    if (EVP_DigestInit(ctx, EVP_sha256()) == 1) {
+        // DigestInit returns 1 on success. But, sigv4 uses 0 as a success
+        // value.
+        ret = 0;
+    } else {
+        ret = -1;
+    }
+
+    return ret;
+}
+
+static int32_t hash_update(void *ctx, const uint8_t *data, size_t data_len) {
+    int32_t ret;
+    ret = EVP_DigestUpdate(ctx, data, data_len);
+
+    if (ret == 1) {
+        // DigestUpdate returns 1 on success. But, sigv4 uses 0 as a success
+        // value.
+        ret = 0;
+    } else {
+        ret = -1;
+    }
+    return ret;
+}
+
+static int32_t hash_final(
+    void *ctx, uint8_t *output_buf, size_t output_buf_len
+) {
+    unsigned int out_buf_len = (unsigned int) output_buf_len;
+
+    int32_t ret = EVP_DigestFinal(ctx, output_buf, &out_buf_len);
+
+    if (ret == 1) {
+        // DigestFinal returns 1 on success. But, sigv4 uses 0 as a success
+        // value.
+        ret = 0;
+    } else {
+        ret = -1;
+    }
+
+    return ret;
+}
+
+size_t aws_sigv4_get_iso8601_time(char buffer[17]) {
+    assert(buffer != NULL);
+    struct timeval tv;
+    struct tm *tm_info;
+
+    gettimeofday(&tv, NULL);
+
+    // NOLINTNEXTLINE(concurrency-mt-unsafe)
+    tm_info = gmtime(&tv.tv_sec);
+
+    return strftime(buffer, 17, "%Y%m%dT%H%M%SZ", tm_info);
+}
+
+GglError aws_sigv4_add_header_for_signing(
+    GglByteVec *vector, GglBuffer header_key, GglBuffer header_value
+) {
+    GglError ret = ggl_byte_vec_append(vector, header_key);
+    ggl_byte_vec_chain_push(&ret, vector, ':');
+    ggl_byte_vec_chain_append(&ret, vector, header_value);
+
+    // Non-canonical delimiter used by the AWS Sigv4 library to separate
```

Comment:
What does non-cannonical mean here? isn't \r\n the normal HTTP separator between headers in a request?

Reply:
I think `\n` is the delimiter used by Sigv4 library when calculating the hash. Thus, this is non-canonical from that perspective.